test/src/tests-batchUploader.ts

```diff
@@ -84,16 +84,603 @@ describe('batch uploader', () => {
+                });
+                const batch3 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 3',
+                });
+
+                // HACK: Directly access uploader to Force an upload
+                const batchesNotUploaded = await (<any>uploader).uploadBatches(
+                    newLogger,
+                    [batch1, batch2, batch3],
+                    false
+                );
+
+                expect(batchesNotUploaded === null).to.equal(true);
+            });
+
+            it('should return batches that fail to unknown HTTP errors', async () => {
+                window.fetchMock.post(
+                    'https://jssdks.mparticle.com/v3/JS/test_key/events',
+                    400
+                );
+
+                const newLogger = new Logger(window.mParticle.config);
+                const mpInstance = window.mParticle.getInstance();
+
+                const uploader = new BatchUploader(mpInstance, 1000);
+
+                const batchValidator = new _BatchValidator();
+
+                const batch1 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 1',
+                });
+
+                const batch2 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 2',
+                });
+                const batch3 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 3',
+                });
+
+                // HACK: Directly access uploader to Force an upload
+                const batchesNotUploaded = await (<any>uploader).uploadBatches(
+                    newLogger,
+                    [batch1, batch2, batch3],
+                    false
+                );
+
+                expect(batchesNotUploaded).to.be.ok;
+
+                expect(
+                    batchesNotUploaded.length,
+                    'Should have 3 uploaded batches'
+                ).to.equal(3);
+
+                expect(
+                    batchesNotUploaded[0].events[0].data.event_name
+                ).to.equal('Test Event 1');
+                expect(
+                    batchesNotUploaded[1].events[0].data.event_name
+                ).to.equal('Test Event 2');
+                expect(
+                    batchesNotUploaded[2].events[0].data.event_name
+                ).to.equal('Test Event 3');
+            });
+        });
+    });
+
+    describe('Upload Workflow', () => {
+        beforeEach(() => {
+            clock = sinon.useFakeTimers({
+                now: new Date().getTime(),
+            });
+        });
+
+        afterEach(() => {
+            window.fetchMock.restore();
+            clock.restore();
+        });
+
+        it('should organize events and batches in the order they are processed and maintain that order when uploading', (done) => {
```

Comment:
nitpick - When I read this test, it only ensures the order of events, not batches.  Each batch is uploaded individually and then the events are checked. The order of the batches aren't ever checked.  You do that in the ensuing tests, not this one.
```suggestion
        it('should organize events in the order they are processed and maintain that order when uploading', (done) => {
```